clippy_lints/src/copy_refcell.rs

```diff
@@ -0,0 +1,112 @@
+use clippy_config::Conf;
+use rustc_hir::{FieldDef, LetStmt, LocalSource};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_middle::ty::layout::TyAndLayout;
+use rustc_session::impl_lint_pass;
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::implements_trait;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects crate-local usage of `RefCell<T>` where `T` implements `Copy`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `RefCell` has to perform extra book-keeping in order to support references, whereas `Cell` does not.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let interior_mutable = std::cell::RefCell::new(0_u8);
+    ///
+    /// *interior_mutable.borrow_mut() = 1;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let interior_mutable = std::cell::Cell::new(0_u8);
+    ///
+    /// interior_mutable.set(1);
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub COPY_REFCELL,
+    perf,
```

Comment:
Unfortunately should be `pedantic`. Passing `Cell` to a function expecting `RefCell` just won't work. This also applies to checking the fields.

Some lints implement a check for the former case, that being using `predicates_of` and `for_each_local_use_after_expr`. This isn't needed though, especially not for a `pedantic` lint.

Reply:
I'll change to pedantic, and a followup can bring this to perf with that added complexity if wanted.